### PR TITLE
Nerf dropped sliders in osu! difficulty calculation

### PR DESCRIPTION
**Pull request by apollo-dw**
In response to the feedback from the recent response to the rework, we're going to nerf plays on slideraim-heavy maps which are very likely to have dropped sliders. This is done via a new attribute, which describes the ratio from the aim difficulty of a map with sliders to the map without sliders. This is used with in conjunction with a formula that estimates dropped slider ends in a play to nerf aim pp!

This proposal also ships a slight nerf to sliders, by increasing `assumed_slider_radius` from `1.65` to `1.8`.

Example scores affected by this: https://www.toptal.com/developers/hastebin/raw/qicelanoza

This is the only intended hotfix for this deployment, any further work will happen next wave.

### CODE REVIEW

**Comment by smoogipoo** [comment]
[Spreasheet](https://docs.google.com/spreadsheets/d/1MGfydNFAuUaBbU5XkUbplob45xQ-yo_aINqbyPRBR6c/edit#gid=694659869) of these changes.

Will update the `pp_all` sheets as I'm able to.

**Review by stanriders — CHANGES_REQUESTED** [review]
Overall looks good to me

File: osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs [hunk 1]
```diff
@@ -125,6 +125,13 @@ private double computeAimValue()
                 aimValue *= 1.0 + 0.04 * (12.0 - Attributes.ApproachRate);
             }
 
+            // We assume 15% of sliders in a map are difficult since there's no way to tell from the performance calculator.
+            double estimateDifficultSliders = Attributes.SliderCount * 0.15;
+            double estimateSliderEndsDropped = Math.Min(estimateDifficultSliders, Math.Min(countOk + countMeh + countMiss, Attributes.MaxCombo - scoreMaxCombo));
+
+            double sliderNerfFactor = (1 - Attributes.SliderFactor) * Math.Pow(1 - estimateSliderEndsDropped / estimateDifficultSliders, 3) + Attributes.SliderFactor;
+            aimValue *= Math.Max(Attributes.SliderFactor, sliderNerfFactor);
```
Comment (MBmasher) [inline osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:133]: I don't understand the purpose of this `Max(Attributes.SliderFactor, sliderNerfFactor)`, does multiplying by `sliderNerfFactor` not yield the same result?

**Review by MBmasher — COMMENTED** [review]
It's good as a quick hotfix

**Comment by MBmasher** [comment]
> [Spreasheet](https://docs.google.com/spreadsheets/d/1MGfydNFAuUaBbU5XkUbplob45xQ-yo_aINqbyPRBR6c/edit#gid=694659869) of these changes.

As far as I understand, this is strictly a nerf, so how are some scores gaining pp?

**Comment by bdach** [comment]
I've checked the score with the biggest relative gain from the spreadsheet (https://osu.ppy.sh/scores/osu/3644047235) and something is off.

2021-11-11-122755_1510x125_scrot

Yep, that's a *negative one* slider dropped. Somehow `Attributes.MaxCombo` is 83, but `scoreMaxCombo` is 84. No idea where to even begin digging with that one, may be easier to clamp it. If I do

```csharp
double estimateSliderEndsDropped = Math.Clamp(Math.Min(countOk + countMeh + countMiss, Attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders);
```

that score gets

```
pp             : 3.5066322747353458
```

which is indeed a nerf (from 3.58104).

File: osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs [hunk 2]
```diff
@@ -125,6 +125,13 @@ private double computeAimValue()
                 aimValue *= 1.0 + 0.04 * (12.0 - Attributes.ApproachRate);
             }
 
+            // We assume 15% of sliders in a map are difficult since there's no way to tell from the performance calculator.
```
[the remark this answers is not in the document]
Comment (apollo-dw) [reply osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:128]: Doable sure - just not sure about using two whole attributes for this. I might just be being overly cautious but I'm under the impression that attribute usage should be very deliberate?

**Comment by apollo-dw** [comment]
> Yep, that's a _negative one_ slider dropped. Somehow `Attributes.MaxCombo` is 83, but `scoreMaxCombo` is 84.

Wow that's a weird one. I think the other buffed scores in the spreadsheet might be ones not recalculated yet that would've gained, but I haven't tested anything just yet.

**Comment by bdach** [comment]
> Wow that's a weird one. I think the other buffed scores in the spreadsheet might be ones not recalculated yet

I... don't think that's how that works? I don't think those scores' PP values are sourced from web, are they?

Anyhow, a few other of the top gaining scores are Big Black SSes, which makes me thing the same sort of thing is happening there.

**Comment by smoogipoo** [comment]
Correct, I recalculate both sr/pp values locally so they represent the final values.

**Comment by smoogipoo** [comment]
Latest changes look to be generating NaN/inf values on many maps (some ranked). E.g. https://osu.ppy.sh/scores/osu/2563062544

Full list of ranked maps:
```
317455506
21559029
47940635
45751059
127749186
2177632918
2616319076
2563062544
```

File: osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (apollo-dw) [reply osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:128]: I tried this as suggested and I didn't think the results were worth another attribute, so will be keeping the estimate. Results provided below (live -> PR -> with new attribute) 
```
https://osu.ppy.sh/scores/osu/3945114629
fieryrage's State of Mind DT: 901pp -> 673pp -> 717pp

https://osu.ppy.sh/scores/osu/3944934181
badeu's Hitogata HDDT: 801pp -> 565pp -> 565pp

https://osu.ppy.sh/scores/osu/3944310727
FGSky's Hitogata HDDTHRNF: 751pp -> 504pp -> 504pp

https://osu.ppy.sh/scores/osu/3944972007
badeu's Lost Umbrella HDHR: 923pp -> 876pp -> 882pp

https://osu.ppy.sh/scores/osu/3944620516
fiery's Dirty Air DT: 649pp -> 610pp -> 620pp

https://osu.ppy.sh/scores/osu/3925637474
Trail Mix's Mugen 666 HDHR: 703pp -> 576pp -> 575pp
```

[the remark this answers is not in the document]
Comment (stanriders) [reply osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:128]: Yeah it's not worth it

**Comment by apollo-dw** [comment]
> Latest changes look to be generating NaN/inf values on many maps (some ranked). E.g. https://osu.ppy.sh/scores/osu/2563062544

Looked to be circle-only maps, should be corrected now.

Score maxcombo and map maxcombo related values should be corrected now too. Any remaining buffs are because of increasing `assumed_slider_radius`, which [Xexxar described in osu!dev](https://discord.com/channels/188630481301012481/380598781432823815/908359760234020875) to be expected.

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 3]
```diff
@@ -34,8 +34,11 @@ protected override DifficultyAttributes CreateDifficultyAttributes(IBeatmap beat
                 return new OsuDifficultyAttributes { Mods = mods, Skills = skills };
 
             double aimRating = Math.Sqrt(skills[0].DifficultyValue()) * difficulty_multiplier;
-            double speedRating = Math.Sqrt(skills[1].DifficultyValue()) * difficulty_multiplier;
-            double flashlightRating = Math.Sqrt(skills[2].DifficultyValue()) * difficulty_multiplier;
+            double aimRatingNoSliders = Math.Sqrt(skills[1].DifficultyValue()) * difficulty_multiplier;
+            double speedRating = Math.Sqrt(skills[2].DifficultyValue()) * difficulty_multiplier;
+            double flashlightRating = Math.Sqrt(skills[3].DifficultyValue()) * difficulty_multiplier;
+
+            double sliderFactor = aimRatingNoSliders / aimRating;
```
Comment (FamousPig) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs:41]: aimRating can be 0(Examples: [aimRatingZero.txt](https://github.com/ppy/osu/files/7523683/aimRatingZero.txt) or the start of https://osu.ppy.sh/beatmapsets/1335452#osu/2766688), so you can get a division by 0 here.

Comment (Wieku) [reply osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs:41]: Noticed the same for https://osu.ppy.sh/beatmapsets/774426#osu/1627673
pp at the first slider jumps to MinInt32

https://user-images.githubusercontent.com/5307179/141386225-6cc8732b-20a0-4447-85ea-69819c38f88a.mp4

Comment (apollo-dw) [reply osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs:41]: Thanks guys, good catch. A bit embarassing that I didn't think of that...

**Comment by smoogipoo** [comment]
Sheet updated with latest changes.

**Comment by proqbr** [comment]
Instead of nerfing the good plays, why not nerf the low combo/low acc plays? That would make a lot more sense and would at the same time be rewarding for the crazy plays like badeu's. This would be an incredibly stupid nerf. Just make it harder to pass bad ranked maps instead, and only nerf low combo plays...

**Comment by Rian8337** [comment]
@proqbr re-read the PR's description. The problem that is being addressed here is not low combo/low acc plays. If you want to address them, it should be done and discussed somewhere or in a separate PR.

Calling this a "useless rework" is also considerably rude. People are trying to improve the system, and this is one of the hundreds of attempts to do so (in fact it's been a long time since we've received an overhaul). A few loopholes here and there are expected as said in the news post. Please respect them instead of throwing a tantrum.

**Review by stanriders — DISMISSED** [review]
Assuming all NaNs are gone, looks good to me

**Review by MBmasher — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]

**Comment by stanriders** [comment]
btw @proqbr Hitogata last diff 98% HD FC pre-nerf was 386pp and post-nerf its 377pp so I'm really not sure where you got your values

**Comment by proqbr** [comment]
> @proqbr re-read the PR's description. The problem that is being addressed here is not low combo/low acc plays. If you want to address them, it should be done and discussed somewhere or in a separate PR.

I am talking about low acc/low combo on slider heavy maps, not in general of course; why would I mention that here. Also my intentions were not to call the rework bad, I'm saying this nerf would make it bad for sure. The very point of the rework is to make plays get their reasonable pp, just because the majority of players are farmers with very limited skill range shouldn't result in such a massive nerf. It was a great step in the right direction, and with this it would go back to not mattering since not only are the old pp maps useless to farm, even these types of maps would end up being useless. The underserved plays all had either bad acc or bad combo (especially nfdt) and only a nerf to such plays was needed. Badeu's much deserved 800pp went to a 500pp that no one will ever care about now; any player would agree to -300pp nerf being too much.

@stanriders I had referred to the spreadsheet and saw 300pp plays on hitogata being nerfed all the way down to 200. My own play on it was a 98 hd about 450 combo play worth 350, and it's completely deserved. The previous 300pps I had were as hard and some even easier as a matter of fact. So nerfing close to FC plays with decent acc doesn't make a lot of sense (and well I guess if anything low acc should be allowed for a dt pass without nf since that's just crazy)...
Also this map definitely deserves 400pp for 99 fc so I hope it's worth that much at least, otherwise doesn't the rework end up being utterly pointless if old farm maps are nerfed and tech maps aren't buffed by much? Pretty much anyone who doesn't spend their entire playtime hitting 1-2 jumps would agree with at least a decent bit of what I mentioned. A sincere request to not go on with the extreme nerf just because of majority disapproval; there are literal people who couldn't even farm the older maps and are still mad - no point in bothering about stuff like that...

**Pull request merged**